tests/spiped_nc/graceful_shutdown.c

```diff
@@ -0,0 +1,96 @@
+#include <signal.h>
+#include <stdlib.h>
+
+#include "events.h"
+#include "warnp.h"
+
+#include "graceful_shutdown.h"
+
+struct graceful_shutdown_cookie {
+	void (* begin_shutdown)(void *);
+	void * caller_cookie;
+};
+
+static volatile sig_atomic_t should_shutdown = 0;
+static void * graceful_shutdown_timer_cookie = NULL;
+
+/* Signal handler for SIGTERM to perform a graceful shutdown. */
+static void
+graceful_shutdown_handler(int signo)
+{
+
+	(void)signo; /* UNUSED */
+	should_shutdown = 1;
+}
+
+/* Requests a graceful shutdown of the caller via the cookie info. */
+static int
+graceful_shutdown(void * cookie)
+{
+	struct graceful_shutdown_cookie * G = cookie;
+
+	/* This timer has expired. */
+	graceful_shutdown_timer_cookie = NULL;
+
+	/* Use the callback function, or schedule another check in 1 second. */
+	if (should_shutdown)
+		G->begin_shutdown(G->caller_cookie);
```

Comment:
`G` is freed in `graceful_shutdown_shutdown()`, which is called by `main.c` in the success and failure (`err4`) cases.